modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsRepositoryImpl.java

```diff
@@ -16,27 +16,32 @@
 package org.apache.ignite.internal.processors.query.stat;
 
 import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.processors.cache.persistence.IgniteCacheDatabaseSharedManager;
+import org.apache.ignite.internal.processors.subscription.GridInternalSubscriptionProcessor;
+import org.apache.ignite.internal.util.typedef.F;
 
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
 
 /**
  * Statistics repository implementation.
  */
 public class IgniteStatisticsRepositoryImpl implements IgniteStatisticsRepository {
     /** Logger. */
-    private IgniteLogger log;
+    private final IgniteLogger log;
 
-    /** Statistics manager. */
-    private final IgniteStatisticsManagerImpl statisticsManager;
+    /** Statistics store. */
+    private final IgniteStatisticsStore store;
 
-    /** Table->Partition->Partition Statistics map, populated only on server nodes without persistence enabled. */
-    private final Map<StatsKey, Map<Integer, ObjectPartitionStatisticsImpl>> partsStats;
+    /** Statistics manager. */
+    private final IgniteStatisticsManagerImpl statisticsMgr;
 
     /** Local (for current node) object statistics. */
-    private final Map<StatsKey, ObjectStatisticsImpl> localStats;
+    private final Map<StatsKey, ObjectStatisticsImpl> locStats;
 
     /** Global (for whole cluster) object statistics. */
     private final Map<StatsKey, ObjectStatisticsImpl> globalStats = new ConcurrentHashMap<>();
```

Comment:
What the difference bw local and global stats?
Does it mean localStats contains only local per-partition raw statistics? or local aggregates as well?
Does it mean globalStats contains only aggregated cluster-wide statistics or raw remote partition stats as well?
If both contains raw partition statistics, can these collections has duplicates?

I thought it make sense to store only raw statistics, and calculate aggregated on deman. Aggregated statistics can cached for performance purposes and thus can be easily invalidated.

Reply:
Partition level statistics contains per-partition statistics, but there are only two difference in statistics objects: partition one have update counter and partId additional fields.
And yes, we store only partition level statistics and aggregate localOne from it and after aggregate global one from each local statistics.
For example: 
node1 partId1 "partition" level statistics: min = 1, max = 10, nulls = 1, total=10, cord = 100, rawCord = byte[]...; updCnt=10, partId=1
node1 part2 "partition" level statistics: min = 11, max = 21, nulls = 1, total=10, cord = 100, rawCord = byte[]...; updCnt=12, partId=2
node2 part3 "partition" level statistics: min = 1, max = 200, nulls = 3, total=10, cord = 100, rawCord = byte[]...; updCnt=19, partId=3
So after aggregation we'll get "local" statistics:
node1 "local" level statistics: min = 1, max = 21, nulls = 2, total=20, cord = 100, rawCord = byte[]...;
node2 "local" level statistics: min = 1, max = 200, nulls = 3, total=10, cord = 100, rawCord = byte[]...;
And after final aggregation to global one:
"global" level statistics: min = 1, max = 200, nulls = 5, total=30, cord = 98, rawCord = byte[]...;